janusgraph-test/src/test/java/org/janusgraph/graphdb/SetEdgeidTest.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    @AfterEach
+    public void shutdown() {
+        graph.close();
+    }
+
+
+    @Test
+    public void graph_add() {
+        initialize();
+
+        JanusGraphTransaction tx = graph.newTransaction();
+        try {
+            tx.addVertex();
+            fail();
+        } catch (Exception ignored) {
+        } finally {
+            tx.rollback();
+        }
+
+        try {
+            // graph.addVertex();
+            System.out.println("graph add element");
+            GraphTraversalSource g = graph.traversal();
+            System.out.println(String.format("26833093692655552l encode result: %s", LongEncoding.encode(26833093692655552l)));
+            System.out.println(String.format("20742965286657360l encode result: %s", LongEncoding.encode(20742965286657360l)));
+            System.out.println("-----------------graph add element addVertex1-----------------");
+            JanusGraphVertex jiulong = graph.addVertex(T.id, 26833093692655552l, T.label, "user", "name", "jiulong");
+            JanusGraphVertex lan = graph.addVertex(T.id, 20742965286657360l, T.label, "user");
+
+            System.out.println("-----------------graph add element addEdge1-----------------");
+            JanusGraphEdge one = jiulong.addEdge("knows", lan, "user", "1", "year", "1993");
+            System.out.println(String.format("first edge id: %s", one.id()));
+
+            System.out.println("-----------------graph add element addEdge2-----------------");
+            JanusGraphEdge two = jiulong.addEdge("get", "knows", lan, "year", "1993");
+            System.out.println(String.format("second edge id: %s", two.id()));
+            // System.out.println(g.E("1-7c7jd2enpog-2dx-5o8rmp86lao").valueMap().toList());
+
+            System.out.println("-----------------graph add element addEdge3-----------------");
+            JanusGraphEdge three = jiulong.addEdge("get", "knows", lan, "year", "1995");
+            System.out.println(String.format("third edge id: %s", three.id()));
+            System.out.println(g.E("get-7c7jd2enpog-2dx-5o8rmp86lao").valueMap().toList());
+
+            System.out.println("-----------------graph add element queryEdge-----------------");
+            System.out.println(g.V().has("name", "jiulong").outE("knows").valueMap().toList());
+            // fail();
+        } catch (Exception e) {
+            System.out.println(e);
+        } finally {
+            graph.tx().rollback();
+        }
+
+    }
+
+    @Test
+    public void traversal_add() {
+        initialize();
+
+        JanusGraphTransaction tx = graph.newTransaction();
+        try {
+            tx.addVertex();
+            fail();
+        } catch (Exception ignored) {
+        } finally {
+            tx.rollback();
+        }
+
+        try {
+            // graph.addVertex();
+            System.out.println("traveral add element");
+            JanusGraphTraversalSource g = graph.JanusTraversal();
```

Comment:
No sure why we need a method `JanusTraversal`. Looks like it is needed only for testing. I would remove that method and would use `new JanusGraphTraversalSource(graph)` here instead.

Reply:
Actually, this class  `JanusGraphTraversalSource` was used for adding the definition of `addE(final String relationId, final String label)`. Because the tinkerpop `GraphTraversalSource` did not have this, so I add this class. And it is not just for testing. When we use traversal obj to add a new edge with customized relationid, should use this class instead of `GraphTraversalSource`.